lib/util.ts

```diff
@@ -325,117 +325,178 @@ export default class Util
     {
         let result: A.Operation = op;
         let doRecursion = true;
+        let copyMetadata = true;
 
         factory = factory || new Factory();
 
         const callback = callbacks[op.type];
-        if (callback)
+        if (callback) {
             // Not sure how to get typing correct for op here
-            ({ result, recurse: doRecursion } = callback(op as any, factory));
+            const recurseResult = callback(op as any, factory);
+            result = recurseResult.result;
+            doRecursion = recurseResult.recurse;
+            copyMetadata = !recurseResult.skipMetadataCopy;
+        }
+
+        const originalMetadata = result.metadata || op.metadata;
+
+        if (!doRecursion) {
+            // Inherit metadata
+            if (originalMetadata && copyMetadata) {
+                result.metadata = originalMetadata;
```

Comment:
What if `result` also contains metadata. Is it preferred behaviour that this gets removed? If not, perhaps `result.metadata = { ...result.metadata, ...originalMetadata }`. Potentially with another check to not have an empty object if both are undefined.

Reply:
Sure, we could do that. The user has a way of overriding this behaviour anyways.